Add: Content lock ability.

jorgefilipecosta opened the pull request:
As shared by @jameskoster in https://github.com/WordPress/gutenberg/pull/42684#issuecomment-1199621547 we are exploring a content lock with the following requirements when a container is 'content locked':

* Non-content child blocks (containers, spacers, columns, etc.) are hidden from the list view, un-clickable on the canvas, and entirely uneditable.
* The Inspector will display a list of all child 'content' blocks. Clicking a block in this list reveals its settings panel. We may need to list the root block as well, but I am not 100% sure.
* The main List View only shows content blocks, all at the same level, regardless of actual nesting.
* Children move and remove is locked.
* Additional child blocks cannot be inserted.
* There is a link in the block toolbar to 'Edit as blocks'. This temporarily toggles off the features outlined above, allowing full edit access, either inline (similar to cropping in the Image block) or in template part focus mode. We should test both approaches.




## Screenshots
![image](https://user-images.githubusercontent.com/11271197/183132649-e20ac31f-bde7-4fd4-83f5-459d192f9f20.png)
![image](https://user-images.githubusercontent.com/11271197/183132683-5fbb410b-8056-4c26-adcf-6f21d065f675.png)
![image](https://user-images.githubusercontent.com/11271197/183133014-d49db0d4-2fe3-405f-b89b-4cb480d1633a.png)

## Sample 
```
<!-- wp:group {"templateLock": "contentOnly", "backgroundColor": "pale-cyan-blue", "layout" :{" type": "default"}} -->
<div class= "wp-block-group has-pale-cyan-blue-background-color has-background"><!-- wp:heading {"style" :{ "typography" :{ "fontStyle": "normal", "fontWeight": "700"}}, "backgroundColor": "pale-pink", "fontSize": "x-large"} -->
<h2 class="has-pale-pink-background-color has-background has-x-large-font-size" style="font-style:normal;font-weight:700">Heading</h2>
<!-- /wp:heading -->

<!-- wp:list -->
<ul><!-- wp:list-item -->
<li>a1</li>
<!-- /wp:list-item -->

<!-- wp:list-item -->
<li>2</li>
<!-- /wp:list-item -->

<!-- wp:list-item -->
<li>3</li>
<!-- /wp:list-item -->

<!-- wp:list-item -->
<li></li>
<!-- /wp:list-item --></ul>
<!-- /wp:list -->

<!-- wp:paragraph {"backgroundColor": "vivid-red", "textColor": "secondary"} -->
<p class="has-secondary-color has-vivid-red-background-color has-text-color has-background">Paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>sdsds</p>
<!-- /wp:paragraph --><cite>sd</cite></blockquote>
<!-- /wp:quote -->

<!-- wp:image {"align": "center", "width":239, "height":239, "sizeSlug": "large", "style" :{ "border" :{ "width": "18px" }}} -->
<figure class="wp-block-image aligncenter size-large is-resized has-custom-border"><img src="https://s.w.org/style/images/about/WordPress-logotype-wmark.png" alt="" style="border-width:18px" width="239" height="239"/><figcaption class="wp-element-caption">f</figcaption></figure>
<!-- /wp:image -->

<!-- wp:spacer {"height":"28px"} -->
<div style="height:28px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->

<!-- wp:group {"backgroundColor":"secondary"} -->
<div class="wp-block-group has-secondary-background-color has-background"><!-- wp:separator -->
<hr class="wp-block-separator has-alpha-channel-opacity"/>
<!-- /wp:separator -->

<!-- wp:spacer {"height":"55px"} -->
<div style="height:55px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->

<!-- wp:pullquote {"gradient":"blush-bordeaux"} -->
<figure class="wp-block-pullquote has-blush-bordeaux-gradient-background has-background"><blockquote><p>End quote inside another group</p></blockquote></figure>
<!-- /wp:pullquote --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

## Testing

- I pasted the previous sample on the code editor.
- I verified there is no way I could select the image (images will be considered content in the future on another PR I'm making) or spacer blocks by clicking on it or by using keyboard navigation, e.g., arrows and tabs. It looks like the non-content block does not exist.
- I verified the non-content blocks don't appear on the list view.
- I verified the group block was totally locked; I could not insert or move any block inside it.
- I verified I could use Edit as blocks to make the block "normal".
- I verified that If I'm using Edit as blocks and move away to another block outside the group, the group block becomes locked again, and the Edit as blocks stops.
- I verified that If I'm using Edit as blocks and click finish editing as blocks, the group block becomes locked again, and the Edit as blocks stops.

## Missing

UI to add a content block. For now, we can hardcode on the code editor. But a UI could be offered either as part of the lock modal or as part of a UI to create patterns on the editor.

jorgefilipecosta committed:
Add: Content lock ability.

jameskoster commented:
Nice, this is working fairly well. A couple of observations:

We need a way to handle non-text content like images. In the example above it should be possible to swap the image source.

It's a bit strange to find the parent container in the 'content' section of the Inspector, since that is not really 'content'. We'll need to adjust the design a bit there.

There's something a little bit awkward about clicking the block in the Inspector and being taken to the settings panel. I think it's because it looks like list view, and so I've come to expect that clicking will just select the block. Having both actions take place simultaneously (select & reveal settings) feels a bit unexpected. Perhaps we need a dedicated 'Settings' button or something. 

---

A more high-level thought that may impact some of the above: 'content lock' kind of suggests to me that only _content_ can be edited. Consequently this brings into question whether it should be possible to adjust the _style_ of the content blocks inside the container? It would simplify things a great deal if you were only able to edit the actual content (and use the tools in the toolbar) here. If it's not a lot of work, it could be interesting to try a version of this where you cannot access the settings panels.

jasmussen commented:
Nice, this is a potent PR, let's be sure to land a variant of this in time for 6.1.

I took it for a quick spin:
![contentlock](https://user-images.githubusercontent.com/1204802/183636949-2ac0a99a-963d-479d-8586-63d770553036.gif)

I'm still absorbing the feature, so I'll need to think more about this. But a few very quick first impressions:

* Locking down and flattening the structure, but maintaining editability of text, is the key value of this.
* The simple logic for the sliding inspector works pretty well for customizing each block inside.
* The flow for _Edit as blocks_ and _finish editing as blocks_ works well, since the unit maintains its overall locking state even if you de-select. In other words, it isn't a "convert to blocks" button. Though, let's look for simpler wording here (Is "Edit" and "Done" too little?)

Something that I'll need to think more about is the heuristic for _which_ blocks show up as still selectable, and which ones don't. I have to edit as blocks to change the image, but I can change the heading level. Mainly this feels a bit confusing as I can clearly see there's an image present, but it doesn't show up in the list view. 

Part of me would like for all of this to be locked down even further, especially for the first iteration. As an unformed thought is to not show _any_ child blocks in list view when locked, and to use a drastically simplified block toolbar for any editable fields inside. Something closer to the caption toolbar we have:
<img width="806" alt="Screenshot 2022-08-09 at 13 42 03" src="https://user-images.githubusercontent.com/1204802/183638935-a5b13091-ff8c-405a-8cfe-0743b749835e.png">

I.e. perhaps instead of this:

<img width="501" alt="Screenshot 2022-08-09 at 13 43 36" src="https://user-images.githubusercontent.com/1204802/183639065-9588c5ac-75b5-4773-8d70-5fa41c2df1e3.png">

maybe something like this:

<img width="606" alt="Screenshot 2022-08-09 at 13 44 07" src="https://user-images.githubusercontent.com/1204802/183639081-8d9796e8-ccd1-4de9-9074-cfb345460ee4.png">

I'll come back with more thoughts, but wanted to share immediate reactions. Nice work!

Mamaduka commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1435,6 +1435,10 @@ export function getTemplateLock( state, rootClientId ) {
 		return state.settings.templateLock;
 	}
 
+	if ( getBlockAttributes( state, rootClientId ).lock?.content === true ) {
+		return 'all';
+	}
+
```
Looking at this change, I wonder if it makes sense for content locking to be a part of `templateLock`, instead of block locking.

jorgefilipecosta replied:
Good question 👍 Yes, I guess we should consider making it a templateLock = content. It seems to make sense.

oandregal replied:
I've been educating myself about the different locking mechanism we have. I agree that using `templateLock` makes more sense than `lock` because `templateLock` aims to control the inner blocks behavior while `lock` controls the block itself.

I see some challenges:

- Naming. We should follow the existing naming schema and indicating _what_ it locks, not what it _allows_ to edit. `noContent`  _could_ work.
- Semantics. Introducing a new state to `templateLock` that is not absorved by the existing `all` is going to create disonance for users of the API. Example:
	- inserts => locks insertion/removal
	- all => "inserts" plus locks movement
	- noContent => "all" plus a bunch of other things (prevent selection for blocks without content, hide them listview, absorb the block inspector of the inner blocks, etc)
- Nested templateLock. The existing locking mechanism fallbacks through a hierarchy: PostType's templateLock < an InnerBlocks's templateLock < another InnerBlocks's templateLock < a singe Block's lock. The latest has more priority. Does the new `noContent` state also flows through the hierarchy? If so, what's the expected behavior here:

```html
<group templateLock=noContent>
  <group templateLock=all />
</group>
```

- Following the reasoning above, there's an existing  `templateLock=false` state. If a block has this state, it "unlocks" anything defined up in the hierarchy. Some blocks use it: social links, navigation, and column. If we respect the cascading behaviour and `false` can unlock anything, the new state won't be effective for those blocks.
- `templateLock` and `lock` and interrelated. We should consider introducing the corresponding states for `lock` if we introduce one for `templateLock`.

Chewing about it, this is the next steps I see:

- Update this PR to use a new state in templateLock instead of lock. Despite the challenges it makes more sense.
- Ignore the value of templateLock for nested containers in this PR. For controlling scope, it should be addressed in a follow-up.
- We may need to consider exposing `templateLock` as an attribute of colum, navigation, and social links should the `false` state needs to be overriden by a pattern. To be tested and investigated when addressing nested content.
 - I'm unsure about introducing a new state for lock. It makes sense conceptually, though I can't see now any practical uses for it. Punt it for the future if need be.

Thoughts?

oandregal replied:
Pushed changes to substitute `lock.content` by `templateLock: "noContent"`. It should work the same as before.

jameskoster commented:
> As an unformed thought is to not show any child blocks in list view when locked

I think this is worth a try, possibly in conjunction with not providing access to the inspector for child blocks as well. These feel somewhat linked to me, and fit the notion of shipping something sooner that is more lightweight.

It would solve the somewhat awkward duplication of child blocks appearing both in list view and the Inspector. And the interaction quirk I pointed out in my last comment, where clicking the child block's name in the Inspector has two detached actions (select + expose settings). It would probably feel more natural if it was just a selection tool. 

> maybe something like this:

That's a good thought. There's really no need to display the block type in the toolbar, especially as transforms shouldn't be possible in this context. Ditto the lock icon – the lock status of these blocks should probably be governed inherently by the parent's 'content lock'. 

Just to connect a dot, could we somehow leverage the work in https://github.com/WordPress/gutenberg/pull/42399# to achieve this?

I would add that it might be worth keeping the parent selector though. It will be helpful in configurations where child blocks fill the height/width of the parent.

jasmussen commented:
> Just to connect a dot, could we somehow leverage the work in https://github.com/WordPress/gutenberg/pull/42399 to achieve this?

I had that thought too. It might, but given how trivial it was in the inspector to hide a few of the toolbar segments, it might be okay to do a temporary thing in the mean time?

> I would add that it might be worth keeping the parent selector though. It will be helpful in configurations where child blocks fill the height/width of the parent.

I don't feel strongly about this. But I do kind of feel like having to select a layer kind of goes against the main purpose of "flattening" a pattern like this. What might be a use case where one such thing was needed in this flattened hierarchy? And might there be a different way to accomplish it?

---

A transversal thought: at the moment a block has options in both the toolbar and in the inspector. I recall @critterverse thinking a lot about this split back in the day, and it remains a bit messy.

As configuration tools in Global Styles → Blocks increasingly get added and mature, it's also becoming clear that if you need to be able to configure a block default, it _has_ to be present in the inspector. It cannot live only in the block toolbar, lest it just won't be fully accessible there. (Because you might be styling a Featured Image on a template that doesn't yet have a featured image inserted in the canvas to select).

This _could_ suggest that every action present in the block toolbar must, somehow, also be present in the inspector. The block toolbar would still afford an immediate and intuitive access point, but nevertheless the option has to be present in the inspector. Which means absorbing controls in an interface like this could potentially be even simpler still. 

Something to think about.

jameskoster commented:
> I don't feel strongly about this. But I do kind of feel like having to select a layer kind of goes against the main purpose of "flattening" a pattern like this.

I don't feel strongly either. It's on my list to test this PR with different block configurations, I think that will help to make a decision.

oandregal commented on packages/blocks/src/store/selectors.js:
```diff
@@ -789,3 +789,13 @@ export const hasChildBlocksWithInserterSupport = ( state, blockName ) => {
 	} );
 };
+
+export const __unstableIsContentBlock = createSelector(
```
What do you think of expanding the existing naming ([example](https://github.com/WordPress/gutenberg/blob/e09fc6ee259385edefc0ae0e14eed82775cba57b/packages/blocks/src/api/utils.js#L293)) and use something along the lines of `__experimentalHasContentRoleAttribute`? As the first time I ran into this, it was not clear what a "content block" was.

oandregal commented on packages/block-editor/src/components/list-view/index.js:
```diff
@@ -63,7 +63,7 @@ function ListView(
 	const { clientIdsTree, draggedClientIds, selectedClientIds } =
 		useListViewClientIds( blocks );
-
+	console.log({clientIdsTree});
```
```suggestion
```

oandregal commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -88,20 +100,35 @@ function ListViewBranch( props ) {
+		},
+		[ parentId ]
+	);
+	const processedBlocks = useMemo( () => {
```
For the ListView component: what do you think of filtering the blocks at this point, so the returned filtered blocks are the ones we want to render. At the moment the logic is spread into ListBranch and ListBlock.

oandregal replied:
Minor: maintaining `filteredBlocks` as a name sounds like something that communicates well its role here.

oandregal replied:
I've pushed these changes. ListViewBlock does not require any knowledge of a "content lock" now :)

oandregal commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -29,19 +39,214 @@ import DefaultStylePicker from '../default-style-picker';
+	);
+}
+
+function BlockNavigatorScreen( { block } ) {
```
What do you think of extracting these components to its own file? The amount of supporting components/utils is high and introduces noise when one wants to understand the high-level flow.

oandregal commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -29,19 +39,214 @@ import DefaultStylePicker from '../default-style-picker';
+									blockTypes={ blockTypes }
+								/>
+							) ) }
+							{ showSelectedBlock && (
```
We can actually still select blocks that don't have content. The question is what we do in that situation. This is what happens now:

https://user-images.githubusercontent.com/583546/184393526-38046555-22f4-45a6-9f56-683b8600e1aa.mp4

My first instinct is that this interaction is at odds with what this feature wants to achieve (hide blocks with no content). So, what do you think of showing the top-level block inspector controls when this happens?

jorgefilipecosta committed:
Update packages/block-editor/src/components/list-view/index.js

Co-authored-by: André <583546+oandregal@users.noreply.github.com>

oandregal commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -29,19 +39,214 @@ import DefaultStylePicker from '../default-style-picker';
+	);
+}
+
+function BlockInspectorAbsorvedBy( { absorvedBy } ) {
```
I've had troubles to parse what `BlockInspectorAbsorvedBy` and `absorvedBy` meant. I'm not sure about the right naming here and wanted to share some ideas to see if we can come up with a better naming schema: `BlockInspectorLockedBlocks` and `topLevelLockedBlock`?

oandregal commented:
Hey, I've reviewed parts of this code (listview, blockinspector) and need to finish it.

In the meantime, I have a conceptual doubt. I've read some of the linked issues and the relationship between this lock and the "block lock" mechanism we have in place is not clear to me. What's the difference between:

![Captura de ecrã de 2022-08-12 13-49-59](https://user-images.githubusercontent.com/583546/184395515-9694a1e2-1bd0-4d6b-8fd2-90e4132e6f53.png)

And this other lock:

![Captura de ecrã de 2022-08-12 13-49-41](https://user-images.githubusercontent.com/583546/184395555-e27a7d5f-6b58-4433-b54d-a0a162df7989.png)

In principle, it sounds like the work here could go into the "lock modal" at a visual level. Conceptually, the lock in this PR feels like a "special/smart" lock that does a few things beyond what the other locks do (affects the inspector and listview) though fall into the boat. Does that make sense? It'd be helpful to finalize the technical review to have more context about these things.

jorgefilipecosta commented:
> Hey, I've reviewed parts of this code (listview, blockinspector) and need to finish it.
> 
> In the meantime, I have a conceptual doubt. I've read some of the linked issues and the relationship between this lock and the "block lock" mechanism we have in place is not clear to me. What's the difference between:
> 
> ![Captura de ecrã de 2022-08-12 13-49-59](https://user-images.githubusercontent.com/583546/184395515-9694a1e2-1bd0-4d6b-8fd2-90e4132e6f53.png)
> 
> And this other lock:
> 
> ![Captura de ecrã de 2022-08-12 13-49-41](https://user-images.githubusercontent.com/583546/184395555-e27a7d5f-6b58-4433-b54d-a0a162df7989.png)
> 
> In principle, it sounds like the work here could go into the "lock modal" at a visual level. Conceptually, the lock in this PR feels like a "special/smart" lock that does a few things beyond what the other locks do (affects the inspector and listview) though fall into the boat. Does that make sense? It'd be helpful to finalize the technical review to have more context about these things.

The other locks allow restrictions on the block like movement or removal, and that's it. This lock is something more advanced. It not only applies restrictions of removal and movement but also tries to make things nested in an area look like a single block, where only the blocks with content are editable.

jorgefilipecosta commented:
Hi @jameskoster, @jasmussen, @oandregal, thank you for the reviews.

> We need a way to handle non-text content like images. In the example above it should be possible to swap the image source.

Images are going to be content when https://github.com/WordPress/gutenberg/pull/43038 is merged.

> I would add that it might be worth keeping the parent selector, though. It will be helpful in configurations where child blocks fill the height/width of the parent.

I would say not the parent because we don't want a hierarchy concept, but on all the toolbars the top block that has the lock would appear as the parent, so we always have a way to select it even if children take up all the space. It would appear as if all the content blocks are direct children no matter what other blocks are in the middle.


So here are the things we should try/next steps.
- Remove the inspector when there is a content lock. Based on this feedback from @jameskoster.
> If it's not a lot of work, it could be interesting to try a version of this where you cannot access the settings panels.
- Correct the labels to "Edit" and "Done".
- Remove child blocks from the list view when the block has content locking.
- Simplify the toolbar and just show some tools. This one is more nuanced e.g: we can not hide almost everything from the toolbar on images as that would make it impossible to replace the image, but on text blocks, we can have a simplified version. 

Does this list of next experiments seems correct, or is there anything I'm missing/should not be tested now?

jasmussen commented:
Sounds good to me.

> Simplify the toolbar and just show some tools. This one is more nuanced e.g: we can not hide almost everything from the toolbar on images as that would make it impossible to replace the image, but on text blocks, we can have a simplified version.

Just to clarify, it would be the exact same categories we remove across all inner blocks, it would not be a curated list. Specifically, it would be the "mover group" and the ellipsis, always. So in the case of paragraphs and images, it would be these sections:
<img width="409" alt="Screenshot 2022-08-15 at 09 45 44" src="https://user-images.githubusercontent.com/1204802/184596457-1ddaae80-aca3-4000-97d1-015b77975f43.png">
<img width="457" alt="Screenshot 2022-08-15 at 09 45 52" src="https://user-images.githubusercontent.com/1204802/184596461-5abc25a9-17a3-4ef5-9080-e3b0d16bd678.png">

oandregal committed:
ListViewBranch: rename back processedBlocks to filteredBlocks

oandregal committed:
ListViewBranch: refactor to make flattenBlockTree function a reducer

oandregal committed:
ListViewBranch: filter the blocks that need rendering

oandregal committed:
ListViewBlock: no longer needs to know anything about content locking

oandregal commented:
I've given this a try using the blockquote that can have innerblocks and also the cite which is a richtext field. Note how:

- the paragraph within the quote is listed in the listview
- the cite within the quote is not listed
- the quote itself is listed

| Unlocked | Locked |
| --- | --- |
| ![Captura de ecrã de 2022-08-16 14-21-17](https://user-images.githubusercontent.com/583546/184879211-9d22478e-95b3-46f0-807b-a3c92b1dd49f.png) | ![Captura de ecrã de 2022-08-16 14-21-24](https://user-images.githubusercontent.com/583546/184879277-131cef83-8695-486e-b4fa-16120f386770.png) |

I'll have a look at what can we do here in the context of Nik's comment about how the nuances of `__experimentalRole` at https://github.com/WordPress/gutenberg/pull/43038#pullrequestreview-1066268642

oandregal commented:
> I've given this a try using the blockquote that can have innerblocks and also the cite which is a richtext field. Note how:

I gave this a shot and it looks like we need to live with this for now. The issue is that the quote's cite is not a block so it's only editable if we make the quote itself editable. If we remove the "role=content" from the quote atts ([see](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/quote/block.json#L24)), there's no way for the user to edit the cite.

oandregal committed:
Remove log

oandregal committed:
Refactor: rename isTemporarlyEditingAsBlocks to isEditingAsBlocks

oandregal committed:
templateLock: remove logic from lock.content

oandregal committed:
templateLock:noContent should prevent drag&drop

oandregal committed:
templateLock:noContent sync innerblocks template if blocks changed

oandregal committed:
templateLock:noContent update docs for InnerBlocks

oandregal committed:
templateLock: substitute lock.content by templateLock to get the locked parent

oandregal committed:
templateLock: substitute lock.content by templateLock in ListView

oandregal committed:
templateLock: substitute lock.content by templateLock in Block toolbar

oandregal committed:
templateLock: add noContent as a new state att in blocks that use it

jameskoster commented:
Would y'all mind updating the testing instructions?

oandregal commented:
@jameskoster I've updated the sample pattern. To be locked, it needs to have `"templateLock":"noContent"` instead of `"lock": {"content": true}`. See [rationale](https://github.com/WordPress/gutenberg/pull/43037#discussion_r941252096).

Edit: ah, also note that since https://github.com/WordPress/gutenberg/pull/43280 the images can be replaced as well.

oandregal committed:
Fix lint issues

jorgefilipecosta committed:
Enhacements

jorgefilipecosta committed:
Remove inspector on content lock

jorgefilipecosta commented:
Thank you a lot for continuing this work, @oandregal 👍 

I added some more updates according to what we discussed:
- Toolbar simplification on children of content-locked blocks.
- Totally removed the inspector control when there is a content lock.
- Correct the labels to "Edit" and "Done".
- Remove child blocks from the list view when the block has content locking.
- Removed the hierarchy concept. The parent selector always shows the block that is locking everything.

Fixed some issues with the change to templateLock, e.g., edit as blocks were not totally unlocking things (besides the attribute we need to update settings).
I also made the content lock inheritance a special case, child blocks should not be able to get out of it; otherwise, things don't work well.

@jameskoster, @jasmussen this implements all the experiments we discussed let me know your thoughts and what we can explore next.

jasmussen commented:
Took it for a spin, here's a quick GIF:
![status](https://user-images.githubusercontent.com/1204802/186371476-01814903-4040-42c4-a125-6fe9e966b27c.gif)

Quick small note: the active state in the inspector isn't legible. We should probably use an active state like either the style toggles or sections in the preferences:

<img width="314" alt="Screenshot 2022-08-24 at 10 37 13" src="https://user-images.githubusercontent.com/1204802/186371673-afaf3dca-2ff8-4db4-b725-998898328152.png">

<img width="292" alt="Screenshot 2022-08-24 at 10 37 20" src="https://user-images.githubusercontent.com/1204802/186371681-25bbb55d-67ec-4012-ab78-d4479a0dca4e.png">

Overall: good progress! The simplification of the toolbar, especially the removal of the ellipsis menu, makes this feel much more considered and intentional.

I still think there's something about that "select parent" button which makes it feel like it's not big enough of a shift to warrant "locking". How sure are we that this button is necessary in most cases? A single group with a full-wide image inside and nothing else, making it go edge to edge in all directions, this would necessitate the button in order to not have to use the list view to select the pattern container. But what other specific examples can you think of that would necessitate the button? I would love to start without this button and add it back only when we find it necessary.

jameskoster commented:
> But what other specific examples can you think of that would necessitate the button?

Your example is the only one I can think of. Somewhat linked to this, do we need to list the container as a clickable item in the Inspector? The presence of the 'Group' item beneath the 'Content' heading feels a bit strange. I agree, we can probably try removing both to begin with.

With images (and other blocks I suppose), should it be possible to change alignment? Or is this verging into 'layout' territory and should therefore be locked?

In the container toolbar 'Edit' seems a bit loose, since it's already editable to some extent. I think 'Edit as blocks' worked a bit better. Perhaps there is other language we can use to describe the functionality? 🤔

jasmussen commented:
> In the container toolbar 'Edit' seems a bit loose, since it's already editable to some extent. I think 'Edit as blocks' worked a bit better. Perhaps there is other language we can use to describe the functionality? 🤔

It is a bit loose, but it's also short and concise, whereas "Edit as blocks" starts to feel complex. How about "Modify"?

jameskoster commented:
Maybe it should be more integrated with the locking UI. Probably something to look into as a follow-up.

jorgefilipecosta committed:
Continued the iterations

jorgefilipecosta commented:
Hi @jasmussen, @jameskoster following the feedback, I applied the following changes:
- Totally remove the parent selector.
- Use the active button state for the content blocks on the inspector.
- Remove the container block from the content blocks on the inspector.
- Changed the label to "Modify".

Regarding things like removing alignment from images, I agree it is something that makes sense. In order to accomplish this, we need changes at the block level, the block is the one that knows that some buttons are supposed to change images, etc, and should be available on a content lock while others are design-related buttons. I guess we can, after merging the content lock mode, have a PR purely adapting our blocks for this mode that can also serve as an example for other blocks in the community if they want to provide the ideal experience.

jameskoster commented:
> I guess we can, after merging the content lock mode, have a PR purely adapting our blocks for this mode that can also serve as an example for other blocks in the community if they want to provide the ideal experience.

I think it's fine to do that as a follow-up, but if we agree that alignment should be disabled in this mode then I would be hesitant about shipping this PR without the follow-up in 6.1.

> Maybe it should be more integrated with the locking UI.

Perhaps the solution is staring directly at us... what if the following lock configuration invoked template lock? 🤔 

<img width="519" alt="Screenshot 2022-08-24 at 17 44 23" src="https://user-images.githubusercontent.com/846565/186475559-90611661-d80c-4974-aefe-6fde6e83a2d2.png">

jasmussen commented:
Nice! This is getting close. Here's a quick status update:
![status](https://user-images.githubusercontent.com/1204802/186627900-04a9ae65-d2f9-4764-9300-27dd3b8f16c5.gif)

This looks good:
<img width="285" alt="Screenshot 2022-08-25 at 11 22 23" src="https://user-images.githubusercontent.com/1204802/186627925-6e584520-4877-4f31-903b-942794f6431a.png">

I think it works well without the parent selector. But right now with it hidden, the positioning of the toolbar is wrong, indented:

<img width="284" alt="Screenshot 2022-08-25 at 11 22 27" src="https://user-images.githubusercontent.com/1204802/186627997-63a711b6-40d0-429c-b3fd-856aafd67dc4.png">

It looks like this CSS is at fault here:

<img width="465" alt="Screenshot 2022-08-25 at 11 24 34" src="https://user-images.githubusercontent.com/1204802/186628140-393a6d22-0440-42df-828d-0f3543343c01.png">

If it's easy to contextually unset that margin in this state, let's do it. If it becomes too ugly and hacky, let's add the parent selector button back and revisit separately, so we can land this first version.

The padding left and right of the "Modify" button (I think the word works okay, we can try it for the first version) is too wide:
<img width="136" alt="Screenshot 2022-08-25 at 11 22 50" src="https://user-images.githubusercontent.com/1204802/186628453-90e14ec2-5472-4f5f-8972-8d3a38ee856b.png">

compare with Replace:

<img width="130" alt="Screenshot 2022-08-25 at 11 23 20" src="https://user-images.githubusercontent.com/1204802/186628491-46b6c76b-4a5b-4a8a-9026-4701d17e8ab7.png">

It looks like this is because the Modify button is nested in an additional wrapping `components-toolbar-group`, which doubles the padding:

<img width="405" alt="Screenshot 2022-08-25 at 11 22 46" src="https://user-images.githubusercontent.com/1204802/186628597-af87a8ec-750a-4152-a621-0f74bc857ec9.png">

Is that an easy fix?

---

> Perhaps the solution is staring directly at us... what if the following lock configuration invoked template lock? 🤔

I do like the idea of fewer new buttons in the toolbar, and embracing existing interfaces. But the template lock modal isn't very clear in this case — and after all you CAN still edit content inside even if parts of this are locked. I wonder what's easier to land in the near term: the "Modify" button, or the modal changes?

jameskoster commented:
> and after all you CAN still edit content inside even if parts of this are locked

Exactly. The modal stipulates that child blocks will be movement/removal locked, which is the exact functionality templatelock exhibits. To be clear, I'm suggesting this as a way to toggle templatelock _on_. 

The 'Modify' button as a way to toggle _off_ seems fine. Although looking at your gif above, if the Inspector wasn't open, it would be quite hard to tell what the result of clicking 'Modify' actually was. That's why I think 'Edit as blocks' worked a bit better despite the verbosity. Not a strong opinion though :)

jorgefilipecosta committed:
Fixed toolbar margin and modify button padding

jorgefilipecosta commented:
HI @jasmussen, @jameskoster,

I updated this PR to fix the toolbar margin and the "Modify" button padding.


Regarding the option in the lock modal vs "Modify" button, I see it as two different things. The modify button locking is not something permanent. It just temporarily allows the user to make changes, but it is not permanently changing the attribute. I guess we can (and should) have an option in the modal to allow adding the content locking state permanently or remove it permanently (can be done in follow-up).


> I think it's fine to do that as a follow-up, but if we agree that alignment should be disabled in this mode, then I would be hesitant about shipping this PR without the follow-up in 6.1.

No issue at all if we agree that some options of the block, like alignment, should be disabled, we can merge this PR, and then the follow-up focused only on block changes can still be created and merged on time for 6.1.

jameskoster commented:
> The modify button locking is not something permanent. It just temporarily allows the user to make changes

Thanks, that's a good distinction to make. I wonder if there's anything we can do to help indicate the temporary nature? Perhaps temporarily invoking spotlight mode on the container?

<img width="1007" alt="Screenshot 2022-08-25 at 11 49 23" src="https://user-images.githubusercontent.com/846565/186645606-12b570e7-3df5-4baf-b256-c6e4a3fae47c.png">

Speaking of spotlight mode, since we want these containers to feel like a single unit, shouldn't the whole thing be focussed when spotlight mode is enabled? Currently only the selected child block is:

<img width="688" alt="Screenshot 2022-08-25 at 11 50 07" src="https://user-images.githubusercontent.com/846565/186645845-fc439771-040e-4fe2-81f1-1d85978423c5.png">

jorgefilipecosta committed:
Fix linting and tests

jasmussen commented:
Thank you for working on this! It feels close!

jorgefilipecosta committed:
Highlight all the block in spotlight mode. Switch to spotlight mode on edit as blocks.

jorgefilipecosta commented:
I made the updates suggested by @jameskoster:
- In spotlight mode, all the block is highlighted.
- When editing as blocks spot light mode is temporarily enabled.

jorgefilipecosta commented:
> Thank you for working on this! It feels close!

Thank you too to you and @jameskoster for suggesting the improvements and helping them get done 👍

jasmussen commented:
Nice, it really is close, all the polish looks good. Current status:
![status](https://user-images.githubusercontent.com/1204802/186876802-7d9ee2e3-e93c-4f36-a3ef-d83372f691ac.gif)

A few comments/questions.

In locked mode, I can't click in the canvas to select the Heading. But I can click in the inspector to change both level, and get full access to deleting the block, etc. — I may have missed a beat, but was this intentional? It seems like you should only get access to the full toolbar and ability to delete/move when you first click "Modify".

Speaking of "Modify" — that still shows the locked-down toolbars. It's my understanding that when modify is toggled, editing the contents inside should be just like editing the contents of a normal group. Is that right?

jorgefilipecosta commented:
Hi @jasmussen,
Thank you for reporting these issues. They were bugs and should now be fixed 👍

jasmussen reviewed (approved):
This is working well:

![content lock](https://user-images.githubusercontent.com/1204802/187615667-24a12d89-c93e-4550-9e3f-7784888be9db.gif)

* The base pattern is simpler to interact with. There's still some multi-selection gnarliness going on, but this is something we can iterate.
* There's still an ability to modify the pattern entirely.

As a strong move forward for patterns, I think we can land a first version of this. 

I would appreciate a code sanity check, though!

jorgefilipecosta committed:
applied feedback

jameskoster commented:
<img width="1434" alt="Screenshot 2022-09-01 at 10 36 39" src="https://user-images.githubusercontent.com/846565/187882941-84f8e06d-0d7e-4c7b-8fae-98b590018187.png">

Did we make a decision here about the active state in list view? In the screenshot above I think I'd expect the Group to be selected. But I wanted to check before opening a ticket :)

jasmussen commented:
Oh right, yes that would be good to follow up on. I would assume the group containing the pattern is selected, but no children are surfaced.

oandregal commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -67,8 +67,15 @@ function ListViewBlock( {
 	const blockInformation = useBlockDisplayInformation( clientId );
-	const blockName = useSelect(
-		( select ) => select( blockEditorStore ).getBlockName( clientId ),
+	const { isContentLocked, blockName } = useSelect(
```
Small PR to get the locking data all in a single hook at https://github.com/WordPress/gutenberg/pull/43803

oandregal commented on packages/block-editor/src/components/list-view/use-block-selection.js:
```diff
@@ -32,7 +32,6 @@ export default function useBlockSelection() {
 	const updateBlockSelection = useCallback(
 		async ( event, clientId, destinationClientId ) => {
 			if ( ! event?.shiftKey ) {
-				await clearSelectedBlock();
```
Why was this change necessary? Tested selection for single and multiple blocks and seems to work fine. Did you run into any issue?

jorgefilipecosta replied:
Hi @oandregal, without this change, if we enable modify as blocks when we select an inner block, the modify as blocks stop. That happens because modifying as blocks stops when the block is unselected, and with this line, there is no block selected for some time.

oandregal commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -109,24 +217,34 @@ const BlockInspector = ( { showNoBlockSelectedMessage = true } ) => {
 	return (
 		<div className="block-editor-block-inspector">
-			<BlockCard { ...blockInformation } />
+			<BlockCard backButton={ backButton } { ...blockInformation } />
```
There are some small leftovers of previous iterations of this PR, such as the back button, that we ended up not using. https://github.com/WordPress/gutenberg/pull/43806 removes them.

oandregal commented on packages/block-editor/src/components/block-card/index.js:
```diff
@@ -8,7 +8,7 @@ import deprecated from '@wordpress/deprecated';
 import BlockIcon from '../block-icon';
 
-function BlockCard( { title, icon, description, blockType } ) {
+function BlockCard( { title, icon, description, blockType, backButton } ) {
```
Removing this unused back button at https://github.com/WordPress/gutenberg/pull/43806

oandregal commented on packages/block-editor/src/hooks/content-lock-ui.js:
```diff
@@ -0,0 +1,161 @@
+			updateBlockAttributes( props.clientId, {
+				templateLock: 'noContent',
+			} );
+			updateBlockListSettings( props.clientId, {
```
I've seen that this updates the `blockListBlockSettings` key in the block-editor store, duplicating the `templateLock` data we already have in `block.attributes`. Why is this necessary?

jorgefilipecosta replied:
Ultimately the source of truth is blockListBlockSettings, the attribute just ends up propagating to blockListBlockSettings. With this change, we avoid a delay.

oandregal commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -125,6 +125,7 @@ Template locking of `InnerBlocks` is similar to [Custom Post Type templates lock
 Template locking allows locking the `InnerBlocks` area for the current template.
 _Options:_
 
```
I've tested inheritance and this is how it works:

- if a CPT [defines](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-templates.md#locking) its `template_lock` to `noContent` will have all "container blocks" (group, cover, column so far) to behave like if all of them have the `templateLock: noContent` applied in their attributes. This is expected and how other values for templateLock work.
- if a individual block defines `templateLock: noContent`, unlike any other locks, no children can override it. I've prepared https://github.com/WordPress/gutenberg/pull/43825 to update the docs.

oandregal replied:
We forgot to document the new state for Custom Post Types. https://github.com/WordPress/gutenberg/pull/43977

mtias commented:
@oandregal @jorgefilipecosta it confused me a lot to see the value being `noContent`, that seems quite obscure. I understand in a sense we are doing the opposite of `lock.content` by only allowing content to be modified. But if that was the concern I'd suggest `contentOnly` or something else that is more clear.

oandregal commented:
Yeah, that was [my concern](https://github.com/WordPress/gutenberg/pull/43037#discussion_r950131396). I've prepared https://github.com/WordPress/gutenberg/pull/44131 to update `noContent` to `contentOnly`. I cannot find a better alternative and I agree `noContent` is confusing.

fabiankaegy commented:
I've just added the Needs Dev Note Label as I think this will be very impactful in WordPress 6.1 :) 

CC: @bph

richtabor commented:
Just to be clear for documentation purposes (as I think the PR name may be a bit misleading) but this mechanism locks everything _but_ content blocks, yea?

And should we refer to this as "Content-only locking"? Not too thrilled with that term myself.

jameskoster commented:
It locks child block position / layout, addition, and removal. Text and attributes like image source remain editable. The objective was to make complex groups feel more like single blocks.

"Content-only editing" might be better? "Group template" could be an option, but "template" is obviously a loaded word already. "Flattened groups"...? Naming is hard.

adamziel commented:
https://github.com/WordPress/gutenberg/pull/44521 makes the new experimental/unstable selectors introduced in this PR private.

fabiankaegy commented:
Out of curiosity, is there a mechanism to limit which user roles have the ability to "Modify" the actual blocks? 

![CleanShot 2022-09-30 at 11 52 50@2x](https://user-images.githubusercontent.com/20684594/193244683-0fb04a7a-af6e-43f3-bd8b-917df5b6966c.png)

I tried to look through the code to see whether it was possible but it didn't seem to be the case. 

I think it would be a great addition to this feature to be able to control which user role levels have the ability to modify the actual block structure / vs which roles can only edit the content :) 

Thanks in advance for any feedback :)

mcsf commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1778,6 +1778,21 @@ export function lastBlockInserted( state = {}, action ) {
 }
 
+/**
+ * Reducer returning the block that is eding temporarily edited as blocks.
```
Typo here. I suppose this was meant to be:

```diff
-Reducer returning the block that is eding temporarily edited as blocks.
+Reducer returning the block that is being temporarily edited as blocks.
```

mcsf commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1778,6 +1778,21 @@ export function lastBlockInserted( state = {}, action ) {
+ *
+ * @return {Object} Updated state.
+ */
+export function temporarilyEditingAsBlocks( state = '', action ) {
+	if ( action.type === 'SET_TEMPORARILY_EDITING_AS_BLOCKS' ) {
+		return action.temporarilyEditingAsBlocks;
```
I think we should improve the nomenclature here. It doesn't convey much information to say that we are editing [a block] _as blocks_. ;) 

This deserves a bit more thought and I don't have real recommendations yet. One thing that crossed my mind was the more explicit `temporarilyUnlockInnerBlocks`.

jasmussen commented:
Created #45263 as a small followup that should be easy.

jorgefilipecosta commented:
> Out of curiosity, is there a mechanism to limit which user roles have the ability to "Modify" the actual blocks?

As referred to previously in a core editor chat, for now, we did not implement role-based restrictions, but it is something we can consider in the future. Thank you for the suggestion 👍
